diary/ads-support.org

```diff
@@ -0,0 +1,62 @@
+#+TITLE: Ads Support
+#+DATE: 2021-01-18
+#+AUTHOR: Zach Mandeville
+
+** Introduction
+This diary is a work-in-progress, documenting how we add ADS support to the xDS
+test harness. It starts by checking assumptions I have about ADS, sets out an
+initial design for the work, and then acts as a work diary of my pgoress.
+
+The bottom of the document has open questions for which I'd love feedback. When
+a question is answered, it'll be removed from that section and its answer woven
+into this diary.
+
+** Assumptions/Current Understanding
+- The xDS transport protocol has [[https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#variants-of-the-xds-transport-protocol][four variants]]. An xDS server can handle any or
+  all of these variants.
+- The test framework, too, should cover all variants.
+  + It [[https://github.com/ii/xds-test-harness/tree/de750b5ba26ba3f0de5a6ecca1fcac20b787ee30][currently]] only supports SoTW(State of the World)/Basic
+  + SoTW/ADS(Aggregated Discovery) should be the added next, followed by Delta/Basic and Delta/ADS.
+- An xDS server is configured for a single variant on startup
+  + e.g: if it's set to SoTW/ADS, then all CDS,LDS,RDS,EDS features will be handled
+    via ADS
+  + It cannot switch itself from basic to ADS and remain running.
+- The intent of our current tests are the same, no matter if a server is
+  configured as basic or aggregated.
+  + However, there will be additional 'ADS-only' tests
+- A team may want to ensure their server can handle basic /and/ ADS.
+- **The framework should be able to be configured for a server's configuration, a configuration that holds through the entire test suite.**
+- **If a team wants to test their server with multiple configurations, they should run the test suite multiple times, once per configuration.**
```

Comment:
Instead of requiring users to run the test suite multiple times, I think it would be better for the test suite to be told which variants of the protocol the xDS server supports, and it should automatically run all of the relevant tests against the server for that set of variants.

One possible way to structure this would be to give each test a list of tags to indicate which variant(s) the test is relevant for.  For example, we can have tags for "SotW", "incremental", "aggregated", and "non-aggregated", and each test can have two or more tags (each test will need at least one of "SotW" or "incremental" and at least one of "aggregated" or "non-aggregated").  For each protocol variant that the server supports, the test framework will go through the list of tests and run all tests that are relevant for that variant -- e.g., when testing non-aggregated SotW, it will run all tests that have both "SotW" and "non-aggregated" tags.